src/edu/washington/escience/myria/operator/CSVFileScanFragment.java

```diff
@@ -200,99 +206,76 @@ protected TupleBatch fetchNextReady() throws IOException, DbException {
-                    source
-                        .getInputStream(adjustedStartByteRange, adjustedStartByteRange + 1)
-                        .read();
-            if (currentChar == '\n') {
-              discardedRecord = true;
-            }
-          }
-        }
-
-        /*
-         * Here, if we are on the last row, we make sure to read the entire row until we either hit a new line or until
-         * we have read the entire file (again, this is for the case where a single worker might be reading a single
-         * large row that was split among other workers)
-         */
-        if (onLastRow && !finishedReadingLastRow && !isLastWorker) {
-          long movingEndByte = adjustedEndByteRange + 1;
-          long characterPositionAtBeginningOfRecord = record.getCharacterPosition();
-          boolean newLineFound = false;
-          while (!newLineFound) {
-            // In order to find the new line character, we use SequenceInputStream to pipe in more bytes
-            InputStream trailingEndInputStream =
-                source.getInputStream(
-                    movingEndByte, Math.min(movingEndByte + byteOverlap, maxByteRange));
-            movingEndByte += byteOverlap;
-            partitionInputStream =
-                new SequenceInputStream(partitionInputStream, trailingEndInputStream);
-            int dataChar = partitionInputStream.read();
-            while (dataChar != -1) {
-              char currentChar = (char) dataChar;
-              if (currentChar == '\n'
-                  || currentChar == '\r'
-                  || Math.min(movingEndByte, maxByteRange) == maxByteRange) {
-                newLineFound = true;
-                // Reading from the source from the beginning again since we only have the character offset information
-                InputStream completePartitionStream =
-                    source.getInputStream(
-                        adjustedStartByteRange, Math.min(movingEndByte, maxByteRange));
-                BufferedReader reader =
-                    new BufferedReader(new InputStreamReader(completePartitionStream));
-                reader.skip(characterPositionAtBeginningOfRecord);
-                parser =
-                    new CSVParser(
-                        reader,
-                        CSVFormat.newFormat(delimiter).withQuote(quote).withEscape(escape),
-                        0,
-                        0);
-                iterator = parser.iterator();
-                record = iterator.next();
-                finishedReadingLastRow = true;
-                break;
-              }
-              dataChar = partitionInputStream.read();
+            if (finalLineFound) {
+              long characterPositionAtBeginningOfRecord = record.getCharacterPosition();
+              InputStream completePartitionStream =
+                  source.getInputStream(
+                      adjustedStartByteRange, Math.min(trailingEndByte, maxByteRange));
```

Comment:
See above for optimization to only fetch the rest of the record instead of the entire `byteOverlap` range.